src/profile-logic/import/art-trace.js

```diff
@@ -0,0 +1,932 @@
+      this._stackTable.data.push([frame, prefix]);
+      this._stackMap.set(key, stack);
+    }
+    return stack;
+  }
+
+  _getOrCreateFrameForMethodId(methodId) {
+    let frame = this._frameMap.get(methodId);
+    if (frame === undefined) {
+      const methodInfo = this._methodMap.get(methodId);
+      let methodString;
+      if (methodInfo === undefined) {
+        // This can happen in practice. The Android Studio profiler shows
+        // 'unknown' in that case.
+        methodString = '<unknown method ID 0x' + methodId.toString(16) + '>';
+      } else {
+        const { className, methodName } = methodInfo;
+        methodString = className + '.' + methodName;
+      }
+      const stringIndex = this._stringTable.length;
+      this._stringTable.push(methodString);
+      const category = this._categoryInfo.inferJavaCategory(methodString);
+      frame = this._frameTable.data.length;
+      this._frameTable.data.push([stringIndex, null, null, null, category]);
+      this._frameMap.set(methodId, frame);
+    }
+    return frame;
+  }
+
+  enterMethod(methodId) {
+    const frame = this._getOrCreateFrameForMethodId(methodId);
+    this._currentStack = this._getOrCreateStack(frame, this._currentStack);
+  }
+
+  exitMethod(_methodId) {
+    if (this._currentStack === null) {
+      // This has been observed to happen in tracing-based traces (rather than sampling-based traces). Not sure why.
+      // console.warn('exiting method when stack is empty');
+      return;
+    }
+    this._currentStack = this._stackTable.data[this._currentStack][1];
+  }
+
+  // Called before enter/exitMethod are called for this time
+  advanceTimeTo(timestampInMSSinceStartTime) {
+    if (this._nextSampleTimestamp === 0) {
+      this._nextSampleTimestamp = timestampInMSSinceStartTime;
+      if (this._name !== 'main') {
+        this._registerTime = timestampInMSSinceStartTime;
+      }
+      return;
+    }
+
+    // Write out the previous sample, and synthesize samples in between the
+    // two timestamps. Unfortunately the ART trace format does not record
+    // timestamps for samples in which the stack does not change. We fill up
+    // such gaps with evenly-spaced synthesized samples at the interval that
+    // was divined earlier.
+    while (this._nextSampleTimestamp < timestampInMSSinceStartTime) {
+      this._samples.data.push([this._currentStack, this._nextSampleTimestamp]);
+      if (this._honorOriginalSamplingTimestamps) {
+        // Only use this loop to fill up any gaps that are at least 2 * interval wide.
+        // When less than 2 * interval remains, snap to the original next timestamp.
+        if (
+          timestampInMSSinceStartTime - this._nextSampleTimestamp <
+          2 * this._intervalInMsec
+        ) {
+          this._nextSampleTimestamp = timestampInMSSinceStartTime;
+          break;
+        }
+      }
+      this._nextSampleTimestamp += this._intervalInMsec;
+    }
+  }
+
+  finish() {
```

Comment:
This method would be worth return an explicit type, to help ensure it doesn't break.

Reply:
Done.